tests/CppWinRTAuthoringTests.cpp

```diff
@@ -162,6 +181,27 @@ TEST_CASE("CppWinRTAuthoringTests::Events", "[property]")
     test.MyTypedEvent.invoke(nullptr, 42);
     test.MyTypedEvent(token2);
 }
+
+TEST_CASE("CppWinRTAuthoringTests::EventsAndCppWinRt", "[property]")
+{
+    struct Test : winrt::implements<Test, winrt::Windows::Foundation::IMemoryBufferReference>

```

Comment:
I initially tried to implement the simpler `INotifyPropertyChanged` interface manually (I know we have helpers for it, but it was simpler to implement), but `untyped_event` and `typed_event` require the event to derive from `` or `` respectively.

I think there are a good number of events that don't derive from either? `Windows.UI.Xaml.Data.CurrentChangingEventHandler`?

FYI @asklar @jonwis

Reply:
True, but those would just be `using=` statements for those folks. The mail thread we had suggested a name like `wil::custom_event<>` as a using? I think that's fine.

Reply:
Could just be
```cpp
template<typename Delegate>
struct custom_event { /* put current definition of event_base here */ };

template<typename EventArgs>
using untyped_event = custom_event<winrt::Windows::Foundation::EventHandler<EventArgs>>;

template<typename Sender, typename EventArgs>
using typed_event = custom_event<winrt::Windows::Foundation::TypedEventHandler<Sender, EventArgs>>;
```

Avoids introducing new types; the `typed_event` and `untyped_event` are just aliases for canned `custom_event` patterns.

Downside is that the name looks ugly in the debugger because the debugger will use the expanded name. (Similar to what you see with `std::wstring` exploding into `std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >`.